File: frontend/src/lib/api/survey.js
import { authClient } from './client';

/**
 * @typedef {{ id: number, text: string, trait: 'O'|'C'|'E'|'A'|'N', reverse: boolean }} Question
 * @typedef {{ score: number, level: 'low'|'mid'|'high' }} TraitScore
 * @typedef {{ openness: TraitScore, conscientiousness: TraitScore, extraversion: TraitScore, agreeableness: TraitScore, neuroticism: TraitScore }} OceanScores
 * @typedef {{ user_id: string, scores: OceanScores, created_at: string, updated_at: string }} PersonalityProfile
 */

/** @returns {Promise<Question[]>} */
export async function getQuestions() {
  const res = await authClient.get('/api/v1/survey/questions');
  return res.data;
}

/**
 * @param {Record<number, number>} answers  questionId → likert value 1-5
 * @returns {Promise<PersonalityProfile>}
 */
export async function submitSurvey(answers) {
  const res = await authClient.post('/api/v1/survey/submit', { answers });
  return res.data;
}

/**
 * Returns the user's saved profile, or null if they haven't taken the survey yet.
 * @returns {Promise<PersonalityProfile|null>}
 */
export async function getMyProfile() {
  try {
    const res = await authClient.get('/api/v1/survey/profile/me');
    return res.data;
  } catch (err) {
    if (err.response?.status === 404) return null;
    throw err;
  }
}

File: Backend/app/schemas/survey.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator


class QuestionOut(BaseModel):
    id: int
    text: str
    trait: str
    reverse: bool


class SurveySubmitIn(BaseModel):
    answers: dict[int, int]

    @field_validator("answers")
    @classmethod
    def validate_answers(cls, v: dict[int, int]) -> dict[int, int]:
        if set(v.keys()) != set(range(1, 45)):
            missing = set(range(1, 45)) - set(v.keys())
            extra = set(v.keys()) - set(range(1, 45))
            parts = []
            if missing:
                parts.append(f"missing ids: {sorted(missing)}")
            if extra:
                parts.append(f"unexpected ids: {sorted(extra)}")
            raise ValueError("; ".join(parts) or "answers must be exactly ids 1–44")
        for q_id, val in v.items():
            if val < 1 or val > 5:
                raise ValueError(f"answer for question {q_id} must be 1–5, got {val}")
        return v


class TraitScore(BaseModel):
    score: float
    level: Literal["low", "mid", "high"]


class OceanScores(BaseModel):
    openness: TraitScore
    conscientiousness: TraitScore
    extraversion: TraitScore
    agreeableness: TraitScore
    neuroticism: TraitScore


class PersonalityProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    scores: OceanScores
    created_at: datetime
    updated_at: datetime
